lib/Backends/CPU/tests/CPUOperatorTest.cpp

```diff
@@ -0,0 +1,146 @@
+      "RowwiseQuantizedSparseLengthsWeightedSum_Float16_AccumFloat16/0",
+      "RowwiseQuantizedSparseLengthsSum_Float16_AccumFloat/0",
+      "RowwiseQuantizedSparseLengthsSum_Float16_AccumFloat16/0",
+      "FusedRowwiseQuantizedSparseLengthsWeightedSum_Float16_AccumFloat/0",
+      "FusedRowwiseQuantizedSparseLengthsWeightedSum_Float16_AccumFloat16/0",
+      "FusedRowwiseQuantizedSparseLengthsWeightedSum_ConvertedFloat16/0",
+      "FusedRowwiseQuantizedSparseLengthsSum_Float16_AccumFloat/0",
+      "FusedRowwiseQuantizedSparseLengthsSum_Float16_AccumFloat16/0",
+      "SparseToDenseMask1/0",
+      "SparseToDenseMask2/0",
+      "FP16Reshape/0",
+      "sliceReshape_Float16/0",
+      "Flatten_Float16Ty/0",
+      "Bucketize/0",
+      "FP16SoftMax/0",
+      "BatchOneHotDataFloat/0",
+      "BatchOneHotDataFloat16/0",
+      "BatchOneHotDataInt64/0",
+      "BatchOneHotDataInt32/0",
+      "BatchOneHotDataInt8/0",
+      "dotProduct1D_Float16/0",
+      "dotProduct2D_Float16/0",
+      "BatchBoxCox_Float/0",
+      "BatchBoxCox_Float16/0",
+      "ConvertFrom_FloatTy_To_Float16Ty/0",
+      "ConvertFrom_FloatTy_To_Int32ITy/0",
+      "ConvertFrom_FloatTy_To_Int64ITy/0",
+      "ConvertFrom_Float16Ty_To_FloatTy/0",
+      "ConvertFrom_Float16Ty_To_Float16Ty/0",
+      "ConvertFrom_Float16Ty_To_Int32ITy/0",
+      "ConvertFrom_Float16Ty_To_Int64ITy/0",
+      "ConvertFrom_Int32ITy_To_FloatTy/0",
+      "ConvertFrom_Int32ITy_To_Float16Ty/0",
+      "ConvertFrom_Int32ITy_To_Int64ITy/0",
+      "ConvertFrom_Int64ITy_To_FloatTy/0",
+      "ConvertFrom_Int64ITy_To_Float16Ty/0",
+      "ConvertFrom_Int64ITy_To_Int32ITy/0",
+      "ConvertFrom_FloatTy_To_Float16Ty_AndBack/0",
+      "ConvertFrom_FloatTy_To_Int32ITy_AndBack/0",
+      "ConvertFrom_FloatTy_To_Int64ITy_AndBack/0",
+      "ConvertFrom_Float16Ty_To_FloatTy_AndBack/0",
+      "ConvertFrom_Float16Ty_To_Float16Ty_AndBack/0",
+      "ConvertFrom_Float16Ty_To_Int32ITy_AndBack/0",
+      "ConvertFrom_Float16Ty_To_Int64ITy_AndBack/0",
+      "ConvertFrom_Int32ITy_To_FloatTy_AndBack/0",
+      "ConvertFrom_Int32ITy_To_Float16Ty_AndBack/0",
+      "ConvertFrom_Int64ITy_To_FloatTy_AndBack/0",
+      "ConvertFrom_Int64ITy_To_Float16Ty_AndBack/0",
+      "ConvertFrom_Int64ITy_To_Int32ITy_AndBack/0",
+      "BasicDivNetFloatVsInt8/0",
+      "BasicAddNetFloatVsFloat16/0",
+      "BasicSubNetFloatVsFloat16/0",
+      "BasicMulNetFloatVsFloat16/0",
+      "BasicDivNetFloatVsFloat16/0",
+      "BasicMaxNetFloatVsFloat16/0",
+      "BasicMinNetFloatVsFloat16/0",
+      "Int16ConvolutionDepth10/0",
+      "Int16ConvolutionDepth8/0",
+      "FP16ConvolutionDepth10/0",
+      "FP16ConvolutionDepth8/0",
+      "FC_Float16/0",
+      "Tanh_Float16/0",
+      "Exp_Float16/0",
+      "rowwiseQuantizedSLWSTest/0",
+  };
+  return blacklist.count(name) == 0;
+}
+
+INSTANTIATE_TEST_CASE_P(CPU, OperatorStatelessTest, ::testing::Values("CPU"));
```

Comment:
This assumes that a backend needs to be aware of all the test suites (not specific tests) so that it can instantiate them. What happens if a new test suite is added in glow/master? This backend wouldn't know, right? 
- May be some kind of a dynamic discovery mechanism could be used to find all test suites and instantiate them dynamically?
- Or e.g. the library containing those test suites should somehow require for each test suite certain symbols to be defined at the linking stage. This way each backend will have to define them (e.g. using `INSTANTIATE_TEST_CASE_P`), because otherwise the linker would complain and make them aware that they miss a new test suite now.

Reply:
Yeah, this is a valid complaint, but as far as I can tell, [the ability to dynamically/programmatically register tests](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#registering-tests-programmatically) is not available in the version of gtest we're using.

Honestly, the number of test suites is not that huge and we already have this problem in that you can forget to add YourBackend to the appropriate INSTANTIATE_TEST_CASE_P.

Reply:
Do you think the static approach (second bullet) would work? E.g. a test suite X defined in a library of tests requires TestSuiteX::static_something to be defined somewhere outside of the library. This way each backend who links against it would need to define it, thus it will be aware of X's existence and can add `INSTANTIATE_TEST_CASE_P`, etc.

Reply:
I like the principle of the static symbol approach, but I don't quite see how it prevents someone from just not creating a test suite for their backend (since the instantiation of the test suite happens in the backend subdirectory, if they just never create it there's nothing to throw an error).

Reply:
> I like the principle of the static symbol approach, but I don't quite see how it prevents someone from just not creating a test suite for their backend (since the instantiation of the test suite happens in the backend subdirectory, if they just never create it there's nothing to throw an error).

I was thinking about something like this:

- Let's say we use a distinct fixture class for each test suite. And such a fixture class defined in the backend-independent library declares a static member of this fixture, but does not define it in any cpp files.  Inside the cpp file with tests, we make sure that this static member is referenced so that the linker would search for it.

- Now any backend inside the backends subdirectory that links against this library, but does not define this static member, would get a linking error, because reference to this static member inside the library cannot be resolved. Once this error pops up (e.g. when a new fixture was added in the library), the backend developer immediately becomes aware of this newly introduced fixture (i.e. about the newly introduced test suite inside the library). Now the required static member can be defined in the backend specific cpp file and `INSTANTIATE_TEST_CASE_P` for this fixture can be added.

Seems like it should work and help backend developers to detect new test suites.

Reply:
> I think I might be able to address your point that it's easy for backends to "overlook" a test suite, too, using a bit of build system sorcery.

Here is a a simpler idea than any build system or linker tricks.

- Let's say we use something like [this](https://github.com/swansontec/map-macro/blob/master/map.h) header file to get a possibility to use the MAP macro over a list of arguments.
- In the test library header file, e.g. `MyTestsLibraryHeader.h`, where the test suites/fixtures are defined,  we just need to define a macro that lists ALL test suites in this library, e.g. like this
```c++
/// List of all test suites defined in the library.
#define TESTSUITE_NAMES TestSuite1, TestSuite2, TestSuite3, TestSuite4
```
- We also define in a common utility header file, e.g. `InstantiateTestSuites.h`, the following macros:
```c++
/// A macro to instantiate a test-suite with a given name using a specific backend.
#define INSTANTIATE_TESTSUITE(NAME) INSTANTIATE_TEST_CASE_P(BACKEND_NAME, NAME, ::testing::Values(STR(BACKEND_NAME)));
/// Stringify the provided macro operand.
#define STR(X) #X

/// Instantiate all test suites from the TESTSUITE_NAMES list using the INSTANTIATE_TESTSUITE macro.
MAP(INSTANTIATE_TESTSUITE, TESTSUITE_NAMES)
```
- Now in the backend-specific cpp file in the backends directory we just need to include the headers mentioned above and then do:
```c++
#define BACKEND_NAME MyCustomBackend
#include "MyTestsLibraryHeader.h"
#include "InstantiateTestSuites.h"
```
- Now we have all test suites instantiated in the backend-specific cpp gtest file! No build system or linker tricks are necessary. Just a bit of a preprocessor magic! :-)

Reply:
Idk man, the build system trick is about 7 lines of CMake :-).  But this is pretty clever, let's see how it compares to what I've got now...